### ISSUE TITLE
Probably wrong implementation for tf.layers.max_pooling1d when data_format='channels_first'

### ISSUE THREAD

**Issue by ZekunZh** [tensorflow/tensorflow#11638]
In function ``call`` of class ``_Pooling1D``,  when the input ``data_format='channels_first'``, it should transform input tensor from '**N,C,H**' to '**N,C,H,W**' (batch_size, channels, height, width), meaning that we should expand dimension on the last dimension. 

However, in the code we use ``inputs = array_ops.expand_dims(inputs, 1)``, expanding on the second dimension and transforming from '**N,C,H**' to '**N,1,C,H**'. Then the ``pool_shape`` and ``strides`` are looking at the third dimension, which is not consistant with our ``expand_dims(inputs, 1)`` used before.

I think the code should be changed to ``inputs = array_ops.expand_dims(inputs, -1)`` and ``return array_ops.squeeze(outputs, -1)``. Using **-1** will expand and squeeze on the last dimension, transforming from '**N,C,H**' to **'**N,C,H,1**', and then doing ``pool_shape`` and ``strides`` on the third dimension.

### Source Code
------------------------
````
  def call(self, inputs):
    # There is no TF op for 1D pooling, hence we make the inputs 4D.
    if self.data_format == 'channels_last':
      inputs = array_ops.expand_dims(inputs, 2)
      pool_shape = (1,) + self.pool_size + (1, 1)
      strides = (1,) + self.strides + (1, 1)
      data_format = 'NHWC'
    else:
      inputs = array_ops.expand_dims(inputs, 1)
      pool_shape = (1, 1) + self.pool_size + (1,)
      strides = (1, 1) + self.strides + (1,)
      data_format = 'NCHW'
````

**Comment by aselle** [comment]
@zhangyaobit, could you take a quick look?

**Comment by zhangyaobit** [comment]
Indeed, looks not right to me. @fchollet, could you take a look?

**Comment by taehoonlee** [comment]
I agree with @ZekunZh. The `expand_dims` in `_Pooling1D` should transform:
- `NHC` to `NHWC ` for `channels_last` with `expand_dims(inputs, 2)`
- `NCH` to `NCHW` for `channels_first` with `expand_dims(inputs, -1)` not `1`

**Comment by zhangyaobit** [comment]
Marked as contribution welcome. Anyone is welcome to implement the fix proposed by @ZekunZh

**Comment by taehoonlee** [comment]
How about contributing your idea @ZekunZh?

**Comment by ZekunZh** [comment]
Thanks for your invitation ! @taehoonlee 
I have proposed my solution in my first post. To be clearer, my suggestion of code for ``call`` function of ``_Pooling1D`` class :
````
  def call(self, inputs):
    # There is no TF op for 1D pooling, hence we make the inputs 4D.
    if self.data_format == 'channels_last':
      ...
    else:
      inputs = array_ops.expand_dims(inputs, -1)
      pool_shape = (1, 1) + self.pool_size + (1,)
      strides = (1, 1) + self.strides + (1,)
      data_format = 'NCHW'

    outputs = ...

    if self.data_format == 'channels_last':
      ...
    else:
      return array_ops.squeeze(outputs, -1)
````

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by taehoonlee** [comment]
The issue is already fixed (refer to the [codes](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/pooling.py#L74)).

**Comment by yongtang** [comment]
Closed through #15500
